test/common/da_hashing.ts

```diff
@@ -1,62 +1,124 @@
-    super();
-    this.nAccountLevel = nAccountLevel;
-    this.nTokenLevel = nTokenLevel;
+  encodedLen(): number {
+    return this.encoder.encodedLen();
   }
 
   encodeTransfer(tx: tx.TranferTx) {
-    this.encodeNumber(Number(tx.from), this.nAccountLevel);
-    this.encodeNumber(Number(tx.to), this.nAccountLevel);
-    this.encodeNumber(Number(tx.tokenID), this.nTokenLevel);
-    this.encodeNumber(Number(encodeFloat(tx.amount)), config.floatLength);
+    this.encoder.encodeNumber(0, 3); //000
+    this.encoder.encodeCommon([tx.from, tx.to, tx.tokenID, tx.tokenID, encodeFloat(tx.amount)], DAEncoder.commonIdx);
+  }
+
+  encodeWithdraw(tx: tx.WithdrawTx) {
+    this.encoder.encodeNumber(2, 3); //010
+    this.encoder.encodeCommon([tx.accountID, tx.accountID, tx.tokenID, tx.tokenID, encodeFloat(tx.amount)], DAEncoder.commonIdx);
   }
 
   encodeDeposit(tx: tx.DepositToOldTx | tx.DepositToNewTx) {
-    this.encodeNumber(Number(tx.accountID), this.nAccountLevel);
-    this.encodeNumber(Number(tx.accountID), this.nAccountLevel);
-    this.encodeNumber(Number(tx.tokenID), this.nTokenLevel);
-    this.encodeNumber(Number(encodeFloat(tx.amount)), config.floatLength);
+    this.encoder.encodeNumber(0, 3); //000
+    this.encoder.encodeCommon([tx.accountID, tx.accountID, tx.tokenID, tx.tokenID, encodeFloat(tx.amount)], DAEncoder.commonIdx);
+  }
+
+  encodeKeyUpdate(tx: tx.DepositToNewTx) {
+    this.encoder.encodeNumber(1, 3); //100
+    assert(tx.amount === BigInt(0));
+    this.encoder.encodeL2Key([tx.accountID, tx.ay], DAEncoder.l2KeyIdx);
   }
 
   encodeNop() {
-    this.encodeNumber(0, txDAEncodeLength(this.nAccountLevel, this.nTokenLevel));
+    this.encoder.encodeNop();
   }
 
-  encodeSpotTrade(tx: tx.SpotTradeTx) {
-    throw new Error('no implementment');
+  encodeSpotTrade(tx: tx.SpotTradeTx, order1: OrderState, order2: OrderState) {
+    let hd = 0;
+    if (order1.isFilled()) {
+      hd |= 2;
+    }
+    if (order2.isFilled()) {
+      hd |= 4;
+    }
+    this.encoder.encodeNumber(hd, 3); //010, 011 or 001
+
+    this.encoder.encodeSpotTrade(
+      [
+        tx.order1AccountID,
+        tx.order2AccountID,
+        tx.tokenID1to2,
+        tx.tokenID2to1,
+        encodeFloat(order1.totalSell()),
+        encodeFloat(order1.totalBuy()),
+        tx.order1Id,
+        encodeFloat(order2.totalSell()),
+        encodeFloat(order2.totalBuy()),
+        tx.order2Id,
+      ],
+      DAEncoder.spotTradeIdx,
+    );
   }
 
   encodeRawTx(tx: tx.RawTx) {
-    const { payload } = tx;
-    this.encodeRawPayload(payload);
+    const { txType, payload } = tx;
+    assert(this.encoder.checkAlign());
+    switch (txType) {
+      case TxType.Deposit:
+        if (payload[TxDetailIdx.Ay1] !== payload[TxDetailIdx.Ay2]) {
+          assert(payload[TxDetailIdx.Amount] === 0n);
+          this.encoder.encodeNumber(1, 3); //100
+          this.encodeRawPayload(payload, 'encodeL2Key');
+          break;
+        }
```

Comment:
do we forget to handle `depositToOld` here?

Reply:
When handling a depositToOld (i.e. txType == Deposit) tx, the code just continue to the next case closure (the one handling TxType.Transfer), because both deposit and transfer use the same encoding schemes ('common') for the payload array.

I would at comment here to clarify the behavior